File: Ефанов Михаил Александрович/Лабораторная работа №1/merge_sort.py
def merge(left, right):
    result = []
    i, j = 0, 0

    # добавляем наименьшие элементы в результирующий массив
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # добавляем оставшиеся элементы в результирующий массив
    while i < len(left):
        result.append(left[i])
        i += 1
    while j < len(right):
        result.append(right[j])
        j += 1
    return result


def merge_sort(arr):
    if len(arr) < 2:
        return arr  # базовый случай
    else:
        middle = int(len(arr) / 2)
        left = merge_sort(arr[:middle])
        right = merge_sort(arr[middle:])
        return merge(left, right)


array = [78, 41, 4, 27, 3, 27, 8, 39, 19, 34, 6, 41, 13, 52, 16]

print("Исходный массив:", array)
print("Отсортированный массив:", merge_sort(array))

File: Тюнибабян Марго/Лабораторная работа №1/mergesort.go
package main

import "fmt"

func merge(left []int,right []int) []int{
	//объединения списков
	lst:=make([]int,0)
	for len(left) > 0 && len(right) > 0{
		if left[0] < right[0]{
			lst = append(lst,left[0])
			left = left[1:]
		}else{
			lst = append(lst,right[0])
			right = right[1:]
		}
	}
	if len(left) > 0{
		lst = append(lst,left...)
	}
	if  len(right) > 0{
		lst = append(lst,right...)
	}

	return lst
}

func mergesort(lst []int) []int{
	//сортировка списка
	length:=len(lst)
	if length >= 2{
		mid := length/2
		lst = merge(mergesort(lst[:mid]),mergesort(lst[mid:]))
	}
	return lst
}

func main(){

	lst := []int{60,-9,32,2,1,876,9,4,11}
	lst = mergesort(lst)
	fmt.Println(lst)
}